File: 450-delete-node-in-a-BST/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func deleteNode(root *TreeNode, key int) *TreeNode {
    if root == nil {
        return nil
    }

    if key > root.Val {
        root.Right = deleteNode(root.Right, key)
    } else if key < root.Val {
        root.Left = deleteNode(root.Left, key)
    } else {
        if root.Left == nil {
            return root.Right
        }
        if root.Right == nil {
            return root.Left
        }

        minNode := findMin(root.Right)
        root.Val = minNode.Val
        root.Right = deleteNode(root.Right, minNode.Val)
    }

    return root
}

func findMin(node *TreeNode) *TreeNode {
    
    for node.Left != nil {
        node = node.Left
    }
    return node
}

File: 450-delete-node-in-a-BST/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:

        def find_min(node: TreeNode) -> TreeNode:
            while node.left:
                node = node.left
            return node
        
        def dfs(root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
            if not root:
                return root

            if key > root.val:
                root.right = dfs(root.right, key)
            elif key < root.val:
                root.left = dfs(root.left, key)
            else:
                if not root.left:
                    return root.right
                if not root.right:
                    return root.left
                
                minNode = find_min(root.right)
                root.val = minNode.val
                root.right = dfs(root.right, minNode.val)
            
            return root
        
        return dfs(root, key)

